hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/api/OMDBInsightEndpoint.java

```diff
@@ -674,6 +685,645 @@ public Response getDeletedDirectorySummary() {
+      Map<String, OmKeyInfo> fsoKeys = searchKeysInFSO(paramInfo);
+      for (Map.Entry<String, OmKeyInfo> entry : fsoKeys.entrySet()) {
+        keysFound = true;
+        KeyEntityInfo keyEntityInfo =
+            createKeyEntityInfoFromOmKeyInfo(entry.getKey(), entry.getValue());
+
+        listKeysResponse.getKeys().add(keyEntityInfo);
+        replicatedTotal += entry.getValue().getReplicatedSize();
+        unreplicatedTotal += entry.getValue().getDataSize();
+      }
+
+      // If no keys were found, return a response indicating that no keys matched
+      if (!keysFound) {
+        return noMatchedKeysResponse(paramInfo.getStartPrefix());
+      }
+
+      // Set the aggregated totals in the response
+      listKeysResponse.setReplicatedDataSize(replicatedTotal);
+      listKeysResponse.setUnReplicatedDataSize(unreplicatedTotal);
+
+      return Response.ok(listKeysResponse).build();
+    } catch (IOException e) {
+      return createInternalServerErrorResponse(
+          "Error searching keys in OM DB: " + e.getMessage());
+    } catch (IllegalArgumentException e) {
+      return createBadRequestResponse(
+          "Invalid startPrefix: " + e.getMessage());
+    }
+  }
+
+  public Map<String, OmKeyInfo> searchKeysInFSO(ParamInfo paramInfo)
+      throws IOException, IllegalArgumentException {
+    Map<String, OmKeyInfo> matchedKeys = new LinkedHashMap<>();
+    // Convert the search prefix to an object path for FSO buckets
+    String startPrefixObjectPath = convertToObjectPath(paramInfo.getStartPrefix());
+    String[] names = parseRequestPath(startPrefixObjectPath);
+    Table<String, OmKeyInfo> fileTable =
+        omMetadataManager.getKeyTable(BucketLayout.FILE_SYSTEM_OPTIMIZED);
+
+    // If names.length > 2, then the search prefix is at the volume or bucket level hence
+    // no need to find parent or extract id's or find subpaths as the fileTable is
+    // suitable for volume and bucket level search
+    if (names.length > 2) {
+      // Fetch the parent ID to search for
+      long parentId = Long.parseLong(names[names.length - 1]);
+
+      // Fetch the nameSpaceSummary for the parent ID
+      NSSummary parentSummary =
+          reconNamespaceSummaryManager.getNSSummary(parentId);
+      if (parentSummary == null) {
+        return matchedKeys;
+      }
+      List<String> subPaths = new ArrayList<>();
+      // Add the initial search prefix object path because it can have both files and subdirectories with files.
+      subPaths.add(startPrefixObjectPath);
+
+      // Recursively gather all subpaths
+      gatherSubPaths(parentId, subPaths, Long.parseLong(names[0]), Long.parseLong(names[1]));
+      // Iterate over the subpaths and retrieve the files
+      for (String subPath : subPaths) {
+        paramInfo.setStartPrefix(subPath);
+        paramInfo.setLimit(paramInfo.getLimit() - matchedKeys.size());
+        matchedKeys.putAll(
+            retrieveKeysFromTable(fileTable, paramInfo));
```

Comment:
Based on limit set in paramInfo, yes will follow sequence for a given subPath first, but it will follow limit restrictions as per below logic:

<img width="882" alt="image" src="https://github.com/apache/ozone/assets/20607992/f0baee1f-aeb2-4295-b3d4-52de248dbcd0">